src/audio/dp_queue.c

```diff
@@ -0,0 +1,386 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+
+#include <sof/audio/dp_queue.h>
+#include <sof/audio/sink_api_implementation.h>
+#include <sof/audio/source_api_implementation.h>
+#include <sof/audio/audio_stream.h>
+
+#include <rtos/alloc.h>
+#include <ipc/topology.h>
+
+struct dp_queue {
+	struct sof_source source_api;  /**< src api handler */
+	struct sof_sink sink_api;      /**< sink api handler */
+
+	uint32_t ibs;
+	uint32_t obs;
+	uint32_t flags;
+
+	size_t data_buffer_size;
+	uint8_t __sparse_cache *data_buffer;
+
+	uint32_t write_offset;
+	uint32_t read_offset;
+	uint32_t available_data; /* amount of data ready for immediate reading */
+	/* free buffer space
+	 * NOTE!
+	 *  - when dp queue is shared, available_data + free_space DOES NOT eq data_buffer_size
+	 *  - when dp queue is not shared available_data + free_space always == data_buffer_size
+	 */
+	uint32_t free_space;
+
+	bool hw_params_configured;
+	struct sof_audio_stream_params audio_stream_params;
+
+	struct k_spinlock lock;
+};
+
+union lock_key {
+	k_spinlock_key_t spinlock;
+	int irq_lock;
+};
```

Comment:
This goes a bit out of the scope of this PR, but given we have very little use of locking in sof/src/audio (mostly in the DAI code, chain-dma and kpb), this is worth a discussion. 

Given we might want to run some of the audio modules in unprivileged code, having code that uses spinlocks or irqlocks is not really future proof. So I wonder if we should just move to mutexes in places like this, which would be available both for user-space and kernel threads.

This does incur a perf penalty, and might cause problems as we still have non-Zephyr SOF users who run audio components in IRQ context, so this is probably not something we can do now, but I'd say this should be the longterm solution for audio code.